test/Microsoft.Azure.Devices.Edge.Test.Common/LeafDevice.cs

```diff
@@ -0,0 +1,352 @@
+
+                    switch (auth)
+                    {
+                        case AuthenticationType.Sas:
+                            return await CreateWithSasAsync(
+                                leafDeviceId,
+                                parentId,
+                                iotHub,
+                                transport,
+                                edgeHostname,
+                                token);
+
+                        case AuthenticationType.CertificateAuthority:
+                            {
+                                string p = parentId.Expect(() => new ArgumentException());
+                                return await CreateWithCaCertAsync(
+                                    leafDeviceId,
+                                    p,
+                                    edgeCa,
+                                    iotHub,
+                                    transport,
+                                    edgeHostname,
+                                    token);
+                            }
+
+                        case AuthenticationType.SelfSigned:
+                            {
+                                string p = parentId.Expect(() => new ArgumentException());
+                                return await CreateWithSelfSignedCertAsync(
+                                    leafDeviceId,
+                                    p,
+                                    useSecondaryCertificate,
+                                    edgeCa,
+                                    iotHub,
+                                    transport,
+                                    edgeHostname,
+                                    token);
+                            }
+
+                        default:
+                            throw new InvalidEnumArgumentException();
+                    }
+                },
+                "Created leaf device '{Device}' on hub '{IotHub}'",
+                leafDeviceId,
+                iotHub.Hostname);
+        }
+
+        static async Task<LeafDevice> CreateWithSasAsync(
+            string leafDeviceId,
+            Option<string> parentId,
+            IotHub iotHub,
+            ITransportSettings transport,
+            string edgeHostname,
+            CancellationToken token)
+        {
+            Device leaf = new Device(leafDeviceId)
+            {
+                Authentication = new AuthenticationMechanism
+                {
+                    Type = AuthenticationType.Sas
+                }
+            };
+
+            await parentId.ForEachAsync(
+                async p =>
+                {
+                    Device edge = await GetEdgeDeviceIdentityAsync(p, iotHub, token);
+                    leaf.Scope = edge.Scope;
+                });
+
+            leaf = await iotHub.CreateDeviceIdentityAsync(leaf, token);
+
+            return await DeleteIdentityIfFailedAsync(
```

Comment:
should it be better to rename "DeleteIdentityIfFailedAsync" to "RunWithErrorHandlingAsync" or "CreateLeafDeviceWithErrorHandlingAsync"?

Reply:
"RunWithErrorHandling" seems too generic. For the second one, I assume you meant "DeleteLeafDeviceWithErrorHandlingAsync", but deleting the leaf device's identity **is** the error handling, so that seems redundant.

I named it to match `CreateDeviceIdentityAsync`, but I can see that I left out the word "Device". I'll add it.

Reply:
The first reason why it looks weird is because its name starts with "Delete" but the calling method is "CreateWithSasAsync".  The intent of this "DeleteIdentityIfFailedAsync" is to do something for failed case and rethrow.  since we can't find a better name, should we add that try-catch block directly in each calling method, to make it more readable?

Reply:
Here's my thinking: the name of the function is `DeleteIdentityIfFailed`. The natural question upon seeing the name is, "Delete the identity if _what_ failed?" To answer that question, you look to the first parameter, and you notice it is a delegate. So you can surmise that the function runs the delegate, and if that fails it will delete the identity.

I can see why a create function with a delete call in it might be jarring. Inlining the code back into the function doesn't get rid of the delete, although it does couch it in try/catch, which is familiar. What it also does however, is reintroduce duplication. I don't think we should delay eliminating duplication until we come up with the perfect name...we give it the most reasonable name we can think of without spending too much time on it, and we let it serve as an annoying (or jarring) reminder that we need to come up with a better name, or more likely, we need to refactor the code to avoid the confusion altogether.

Reply:
This seems a test resource clean up issue; would it be handled by using IDisposable and/or test setup/tear down?

Reply:
It's not a setup/teardown issue since the identity depends on arguments to the test (you can't pass those arguments to the setup routine). I looked at making LeafDevice disposable early on, but Dispose isn't async so that's annoying.